Rectangular rotated surface code

francois-marie opened the pull request:
Adding the possibility to simulate rectangular rotated surface code by adding `dX`.
It defaults to `None` so that the circuit is squared.
The data qubits are placed accordingly (coordinates and indices).

francois-marie committed:
add dX to CircuitGenParameters field

francois-marie committed:
add dX and dZ to rotated surface code

francois-marie committed:
add dX to generate_circuit

francois-marie committed:
test square surface code with dX specified

oscarhiggott commented:
Hi @francois-marie, thank you for the contribution! And apologies for the slow response.

Your changes look good, though I have a couple of proposed further changes before merging. Also, I just pushed some local changes I'd made which added unrotated toric codes and some more tests, so there are some conflicts that need to be resolved unfortunately.

I think it would be a bit clearer if `dX` was renamed to `x_distance` and if `dZ` was renamed to `z_distance`, and if a new `z_distance` parameter was added rather than reusing `distance`. E.g. so that `CircuitGenParams` became:

```
@dataclass
class CircuitGenParameters:
    code_name: str
    task: str
    rounds: int
    distance: int = None
    x_distance: int = None
    z_distance: int = None
    after_clifford_depolarization: float = 0
    before_round_data_depolarization: float = 0
    before_measure_flip_probability: float = 0
    after_reset_flip_probability: float = 0
    exclude_other_basis_detectors: bool = False

```

and then with `generate_circuit` becoming:

```
def generate_circuit(
    code_task: str,
    *,
    rounds: int,
    distance: int = None,
    x_distance: int = None,
    z_distance: int = None,
    after_clifford_depolarization: float = 0.0,
    before_round_data_depolarization: float = 0.0,
    before_measure_flip_probability: float = 0.0,
    after_reset_flip_probability: float = 0.0,
    exclude_other_basis_detectors: bool = False
) -> stim.Circuit:
```

Then there would need to be a check that either `distance` is not None or that both `x_distance` and `z_distance` are not `None`, raising a `ValueError` if not.  Would also need to raise a `NotImplementedError` if `x_distance` and `z_distance` are supplied to generate a circuit that they're not yet supported for.

Finally, you'll see on the master branch now there's a [`test_generated_circuit_graphlike_distance`](https://github.com/oscarhiggott/StimCircuits/blob/bce2818d3a53d3a14f301bcb324719a20fa752d1/stimcircuits/test_surface_code.py#L88) test. So once you've merged master into your own branch it would be nice to add some rectangular cases to that test to verify in both x and z basis.

Let me know if any of this is unclear, and thanks again for the contribution!

francois-marie committed:
dX renamed to x_distance and dZ renamed to z_distance, new z_distance parameter added rather than using distance

francois-marie committed:
remove previous test

francois-marie committed:
check that either distance is not None or that both x_distance and z_distance are not None, raising a ValueError if not

francois-marie committed:
raise a NotImplementedError if x_distance and z_distance are supplied to generate a circuit that they're not yet supported for

francois-marie committed:
check for minimal value of distance if not None or x_distance and z_distance if rectangular

francois-marie committed:
test generated circuit graphlike distance with distance x_distance and z_distance supplied

francois-marie committed:
test generated rectangular circuit graphlike distance

francois-marie committed:
test not implemented rectangular circuit for unrotated surface code and toric code

francois-marie commented:
Hi @oscarhiggott!

I am really sorry for the late response!
Thank you very much for your complete answer and well done for adding unrotated toric codes, this looks really nice!
I tried by best to implement your proposed changes:

- `dX` was renamed to `x_distance` and `dZ` was renamed to `z_distance`, and a new `z_distance` parameter was added rather than reusing `distance`
- There is a check that either `distance` is not `None` or that both `x_distance` and `z_distance` are not `None`, raising a `ValueError` if not
- `generate_surface_or_toric_code_circuit_from_params` raises a `NotImplementedError` if `x_distance` and `z_distance` are supplied to generate a circuit that they're not yet supported for
- I have added a `test_generated_circuit_with_distances_graphlike_distance` test for the case where `distance`, `x_distance` and `z_distance` are supplied. It may be useless, tell me if you would like me to remove it
- I have added a `test_generated_rectangular_circuit_graphlike_distance` test for rotated surface code that add some rectangular cases to verify in both x and z basis. Would you prefer to merge it to `test_generated_circuit_graphlike_distance` directly ?
- Finally, I have added a `test_not_implemented_rectangular_circuit` to check that it raises the correct exception. Once again, it may be useless, tell me if you would also like me to remove it


Tell me if this corresponds to what you had in mind or what can be done to help !

oscarhiggott commented:
This looks great, thank you for the contribution! Apologies for the long delay reviewing and merging it
